src/crontab.c

```diff
@@ -479,16 +479,78 @@ static void delete_cmd(void) {
 		fprintf(stderr, "path too long\n");
 		exit(ERROR_EXIT);
 	}
+	
+	if(backup_crontab(n) == -1){
+		fprintf(stderr, "no crontab for %s\n", User);
+		exit(ERROR_EXIT);
+	}
+
 	if (unlink(n) != 0) {
-		if (errno == ENOENT)
-			fprintf(stderr, "no crontab for %s\n", User);
-		else
-			perror(n);
+		perror(n);
 		exit(ERROR_EXIT);
 	}
 	poke_daemon();
 }
 
+/* returns	0	on success
+ *		-1	on non existent crontab file
+ */
+static int backup_crontab(char crontab_filepath[]) {
+	const char *env_value;
+	char backup_filename[MAX_FNAME];
+	
+	if(env_value = getenv("XDG_CACHE_HOME")){
+		if (!glue_strings(backup_filename, sizeof backup_filename, env_value,
+			"crontab.bak", '/')) {
+				fprintf(stderr, "path too long\n");
+				exit(ERROR_EXIT);
+		}
+	}else if(env_value = getenv("HOME")){
+		if (!glue_strings(backup_filename, sizeof backup_filename, env_value,
+			".cache/crontab.bak", '/')) {
+				fprintf(stderr, "path too long\n");
+				exit(ERROR_EXIT);
+		}
+	}else{
+		fprintf(stderr, "Could not find environment variable XDG_CACHE_HOME or HOME to save the backup\n");
+		exit(ERROR_EXIT);
+	}
+
+	FILE *crontab_file;
+	FILE *backup_file;
+
+	if (!(crontab_file = fopen(crontab_filepath, "r"))) {
+		if (errno != ENOENT) {
+			perror(crontab_filepath);
+			exit(ERROR_EXIT);
+		}
+		return -1;
+	}
+	
+	if (swap_uids() == -1) {
```

Comment:
I wonder, how this will work if `crontab -u user` is invoked by root. One option would be to disable the backup processing if `-u` is used.

Reply:
> * (if so, we could always add the user's name to the backup file)

With this amendment I'd be OK with it. Of course add the username only if -u is used.

Reply:
Will do ! Is there a reason in particular to only do this with -u ?

Reply:
Since multiple files might be written, I also created a directory for crontab